avocado/utils/vmimage.py

```diff
@@ -0,0 +1,375 @@
+        image = self.image.format(version=version,
+                                  build=self.args['build'],
+                                  arch=self.args['arch'])
+        parser = VMImageHtmlParser(image)
+        try:
+            content = urllib2.urlopen(url_images).read()
+            parser.feed(content)
+        except:
+            raise ImageProviderError('Cannot open %s' % url_images)
+        if parser.versions:
+            return url_images + max(parser.versions)
+        else:
+            raise ImageProviderError("No images matching '%s' at '%s'" %
+                                     (image, url_images))
+
+    def remove(self):
+        """
+        Removes the VM Image
+        """
+        if self.path is not None:
+            os.remove(self.path)
+            self.path = None
+
+    @staticmethod
+    def _create_backed_image(source_image):
+        """
+        Creates a Qcow2 backed image using the original image
+        as backing file.
+        """
+        qemu_img = utils_path.find_command('qemu-img')
+
+        name, extension = os.path.splitext(source_image)
+        new_image = '%s-%s%s' % (name, str(uuid.uuid4()).split('-')[0],
+                                 extension)
+        cmd = '%s create -f qcow2 -b %s %s' % (qemu_img,
+                                               source_image,
+                                               new_image)
+        process.run(cmd)
+        return new_image
+
+    @staticmethod
+    def _extract(path, force=False):
+        """
+        Extracts a XZ compressed file to the same directory.
+        """
+        extracted_file = os.path.splitext(path)[0]
+        if not force and os.path.exists(extracted_file):
+            return extracted_file
+        with open(path, 'r') as file_obj:
+            with open(extracted_file, 'wb') as newfile_obj:
+                newfile_obj.write(lzma.decompress(file_obj.read()))
+        return extracted_file
+
+
+class FedoraImageProvider(ImageBase):
+
+    name = 'Fedora'
+
+    def __init__(self, **kwargs):
+        super(FedoraImageProvider, self).__init__(**kwargs)
+        # Here you can use string formatting tags corresponding to the
+        # arguments accepted (version, build and arch).
+        self.url_versions = 'https://dl.fedoraproject.org/pub/fedora/linux/releases/'
+        self.url_images = self.url_versions + '{version}/CloudImages/{arch}/images/'
+        self.image = 'Fedora-Cloud-Base-{version}-{build}.{arch}.qcow2'
+
+        # Set the defaults to be used when the arguments are
+        # not provided. If you don't want to force a version,
+        # you can provide a regular expression that matches the
+        # corresponding format and we will get the highest
+        # occurrence of that expression.
+        self.default_version = '[0-9]+'  # 22/23/24/...
+        self.default_build = '[0-9]+.[0-9]+'  # 1.4/5.1/5.2/...
+        self.default_arch = 'x86_64'
```

Comment:
default arch should also be queryable and not hardcoded.

Reply:
and perhaps even based on the current architecture.

Reply:
Actually I'd simply use the current architecture as default and raise exception when that architecture is not available. I believe it'd be easier to use.

Reply:
Ack. Current system arch it is.